### PR TITLE
Reenable C4244 in repo

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
In response to https://github.com/dotnet/runtime/pull/69236

@lateralusX @lambdageek Do you know off the top of your heads the flag we can use to enable the warning for clang/gcc?

### CODE REVIEW

**Comment by lambdageek** [comment]
@AaronRobinsonMSFT I'm not sure there's anything that is precisely the same as C4244, but I think we previously talked about `-Werror=shorten-64-to-32` and `-Werror=implicit-int-conversion`

They would go near here:
https://github.com/dotnet/runtime/blob/7d864e3f972e2ffe2446c3cb1a1f60e156379deb/src/mono/CMakeLists.txt#L509-L511

**Review by lambdageek — APPROVED** [review]

**Comment by lambdageek** [comment]
@AaronRobinsonMSFT I got a ton of warnings for `shorten-64-to-32` I'm working on a PR to fix it up.  Should have something in a couple of hours

**Comment by AaronRobinsonMSFT** [comment]
> @AaronRobinsonMSFT I got a ton of warnings for `shorten-64-to-32` I'm working on a PR to fix it up. Should have something in a couple of hours

Oh nice. I fixed a bunch of issues for `-Werror=implicit-int-conversion`. Only ran on arm64 so probably missed some. I focus on those if you can handle the others.

**Comment by lateralusX** [comment]
One comment, I didn't reenable 4244 yet since there will be additional issues that needs to be fixed in the cross compiler builds (include some arm32/64 source files as well as some WASM specific source files) that I through would made sense to fix locally first, since doing that through CI is rather time consuming.

**Comment by lateralusX** [comment]
@AaronRobinsonMSFT @lambdageek Maybe we need to coordinate with https://github.com/dotnet/runtime/pull/70062, looks like they are fixing the same issues in some scenarios.

**Comment by lateralusX** [comment]
Maybe rename this PR since it now seems to cover several CoreCLR sources as well?

**Comment by AaronRobinsonMSFT** [comment]
@lambdageek and @lateralusX I've reverted the change that turned "implicit-int-conversion" into an error. This change was proving to be far to impacting - particularly in the .NET PAL layer. I'm not convinced there is a need for this work for SDL so am keeping the enabling of 4244 on Windows builds and keeping the various places that I updated. We can reconsider adding back the error post-.NET 7.

**Comment by AaronRobinsonMSFT** [comment]
@lambdageek or @lateralusX Please take another look here. The PR is green without the "implicit-int-conversion" as error. This work satisfies 4244 for SDL.

File: src/mono/mono/mini/mini-llvm.c [hunk 1]
```diff
@@ -1983,13 +1983,13 @@ get_aotconst_name (MonoJumpInfoType type, gconstpointer data, int got_offset)
...
 		for (size_t i = 0; i < len; ++i)
-			name [i] = tolower (name [i]);
+			name [i] = (char) tolower (name [i]);
 		break;
 	default:
 		name = g_strdup_printf ("%s_%d", mono_ji_type_to_string (type), got_offset);
 		len = strlen (name);
 		for (size_t i = 0; i < len; ++i)
-			name [i] = tolower (name [i]);
+			name [i] = (char) tolower (name [i]);
```
Comment (lateralusX) [inline src/mono/mono/mini/mini-llvm.c:1992]: GINT_TO_CHAR

**Review by lateralusX — COMMENTED** [review]
Mainly looked at the Mono bits, a couple of smaller comments.

**Review by lateralusX — APPROVED** [review]
Mono bits LGTM!

**Comment by AaronRobinsonMSFT** [comment]
@janvorli Can you give the PAL changes a quick once over? Almost all of them are casts.

**Review by janvorli — APPROVED** [review]
PAL changes LGTM. But I cannot see the `-Werror=implicit-int-conversion` being added in this PR. Is it intentional?

**Comment by AaronRobinsonMSFT** [comment]
> PAL changes LGTM. But I cannot see the `-Werror=implicit-int-conversion` being added in this PR. Is it intentional?

I dropped adding it as it started to trigger on other things in the PAL that were giving me pause. I reverted the link and described why at https://github.com/dotnet/runtime/pull/70026#issuecomment-1150543114. The last few I found were related to changes in the exception unwind data structures.

The priority for this work was for enabling 4244 for SDL requirements, but we should revisit as we work to enable other hygiene checks.

**Comment by janvorli** [comment]
Ok, makes sense.

**Pull request merged**